build.gradle

```diff
@@ -24,6 +24,28 @@ subprojects {
                 languageVersion = of(8)
             }
         }
+
+        if (parent == rootProject) {
+            // Add a META-INF/pluginVersion.properties resource into each plugin jar.
+            def generateResources = tasks.register('generateResources') {
+                def versionFile = layout.buildDirectory.dir('generated').map { dir ->
```

Comment:
wouldn't it be better to use 
```
project.convention.getPlugin(JavaPluginConvention.class)
    .sourceSets
    .getByName(SourceSet.MAIN_SOURCE_SET_NAME)
    .resources
    .outputDir
```
?

Reply:
Better how? :thinking:

I've actually based this solution on [this advice](https://discuss.gradle.org/t/how-can-a-custom-gradle-plugin-determine-its-own-version/36761/3) from the Gradle Forums.

Reply:
I had 2 concerns:
1 - Gradle changes the location of the generated source and the `generated` directory doesn't exist anymore at some point (and I think code would break in that case)
2 - The project compiles with an annotation processors that, for some reason, creates an `pluginVersion.properties` that you would be overwriting here; this is not possible though because the actual destination of generated code is not the `generated` folder directly but some more nested path inside it, so this is fine

Reply:
I think that's the point - I create _my own_ output directory which has its own dedicated task to generate its contents. Gradle will create the directory automatically, and will also complain if the final jar tries to contain two files called `META-INF/pluginVersion.properties`.

[Gradle 7 is becoming more picky about such things.](https://discuss.gradle.org/t/gradle-7-0-seems-to-take-an-overzealous-approach-to-inter-task-dependencies/39656/2)